engine-tests/src/test/java/org/terasology/world/chunks/localChunkProvider/LocalChunkProviderTest.java

```diff
@@ -0,0 +1,314 @@
+        final EntityRef mockEntity = mock(EntityRef.class);
+        when(entityManager.create()).thenReturn(mockEntity);
+
+        chunkProvider.completeUpdate();
+
+        verify(mockEntity).addComponent(eq(testComponent));
+    }
+
+    @Test
+    public void testCompleteUpdateGeneratesStoredEntitiesFromPrefab() throws Exception {
+        final Chunk chunk = mockChunkAt(0, 0, 0);
+        final Prefab prefab = mock(Prefab.class);
+        final ChunkProviderTestComponent testComponent = new ChunkProviderTestComponent();
+        final EntityStore entityStore = createEntityStoreWithPrefabAndComponents(prefab, testComponent);
+        final List<EntityStore> entityStores = Collections.singletonList(entityStore);
+        final ReadyChunkInfo readyChunkInfo = ReadyChunkInfo.createForNewChunk(chunk, new TShortObjectHashMap<>(), entityStores);
+        when(chunkFinalizer.completeFinalization()).thenReturn(readyChunkInfo);
+        final EntityRef mockEntity = mock(EntityRef.class);
+        when(entityManager.create(any(Prefab.class))).thenReturn(mockEntity);
+
+        chunkProvider.completeUpdate();
+
+        verify(entityManager).create(eq(prefab));
+        verify(mockEntity).addComponent(eq(testComponent));
+    }
+
+
+    @Test
+    public void testCompleteUpdateRestoresEntitiesForRestoredChunks() throws Exception {
+        final Chunk chunk = mockChunkAt(0, 0, 0);
+        final ChunkStore chunkStore = mock(ChunkStore.class);
+        final ReadyChunkInfo readyChunkInfo = ReadyChunkInfo.createForRestoredChunk(chunk, new TShortObjectHashMap<>(), chunkStore, Collections.emptyList());
+        when(chunkFinalizer.completeFinalization()).thenReturn(readyChunkInfo);
+
+        chunkProvider.completeUpdate();
+
+        verify(chunkStore).restoreEntities();
+    }
+
+    @Test
+    public void testCompleteUpdateSendsBlockAddedEvents() throws Exception {
+        final Chunk chunk = mockChunkAt(0, 0, 0);
+        final short blockId = 42;
+        final EntityRef blockEntity = mock(EntityRef.class);
+        registerBlockWithIdAndEntity(blockId, blockEntity, blockManager);
+        final TShortObjectHashMap<TIntList> blockPositionMappings = new TShortObjectHashMap<>();
+        blockPositionMappings.put(blockId, withPositions(new Vector3i(1, 2, 3)));
+        final ReadyChunkInfo readyChunkInfo = ReadyChunkInfo.createForRestoredChunk(chunk, blockPositionMappings, mock(ChunkStore.class), Collections.emptyList());
+        when(chunkFinalizer.completeFinalization()).thenReturn(readyChunkInfo);
+
+        chunkProvider.completeUpdate();
+
+        final ArgumentCaptor<Event> eventArgumentCaptor = ArgumentCaptor.forClass(Event.class);
+        verify(blockEntity, atLeastOnce()).send(eventArgumentCaptor.capture());
+        final Event event = eventArgumentCaptor.getAllValues().get(0);
+        assertThat(event, instanceOf(OnAddedBlocks.class));
+        assertThat(((OnAddedBlocks) event).getBlockPositions(), hasItem(new Vector3i(1, 2, 3)));
```

Comment:
Maybe a TODO comment here would be good for indicating that it is a bit strange to have an OnAddedBlocks event for a restored chunk.